File: sdk/ts/src/output/formatter.ts
/**
 * @module output/formatter
 *
 * Core Formatter interface + Options/OptionSpec/ColumnSpec types and
 * parseOptions helper. Mirrors Go's go/console/output Formatter interface.
 */

/** Kinds of values an OptionSpec accepts. */
export type OptionType = 'string' | 'int' | 'bool' | 'enum';

/** Describes one option accepted by a Formatter via --format-opt key=value. */
export interface OptionSpec {
  readonly name: string;
  readonly type: OptionType;
  readonly default?: string | number | boolean;
  readonly usage: string;
  readonly enum?: readonly string[];
}

/**
 * Validated map of option values produced by parseOptions. Keys are
 * OptionSpec.name; values are coerced to the spec's declared type.
 */
export type Options = Readonly<Record<string, string | number | boolean>>;

/**
 * Column metadata used by table/csv/text formatters and --cols validation.
 *
 * - `header` is the user-visible label and the value matched against --cols.
 * - `key` is the property name on the row object.
 * - `priority` (optional) controls hide-on-overflow for table; higher wins.
 */
export interface ColumnSpec {
  readonly header: string;
  readonly key: string;
  readonly priority?: number;
}

/**
 * A Formatter encodes structured data to a writable stream.
 *
 * Implementations declare their key, file extensions, and the option keys
 * they accept. The dispatch helper validates --format-opt input against
 * options() before invoking render(), so render() may trust opts to only
 * contain declared keys with values coerced to declared types.
 */
export interface Formatter<T = unknown> {
  /** Unique format identifier exposed via --format <key>. */
  readonly key: string;

  /**
   * File extensions (with leading dot, e.g. ".csv") that map to this
   * formatter for --output extension inference. May be empty.
   */
  readonly extensions: readonly string[];

  /** Option specs accepted by this formatter via --format-opt key=value. */
  readonly options: readonly OptionSpec[];

  /**
   * Renders data to out.
   *
   * @param out Destination writable stream.
   * @param data Single row or readonly array of rows.
   * @param opts Validated options; only declared keys with coerced values.
   * @param cols User-requested column projection; empty means "all columns".
   */
  render(
    out: NodeJS.WritableStream,
    data: T | readonly T[],
    opts: Options,
    cols: readonly string[],
  ): Promise<void> | void;
}

/**
 * Validates raw `key=value` pairs against specs and returns a coerced
 * Options map. Unknown keys, type errors, and out-of-enum values throw an
 * Error listing the offending key and the valid set.
 *
 * A pair without `=` (e.g. `"no-header"`) is treated as bool true; only
 * valid when the matching spec has type `'bool'`.
 *
 * Defaults from specs fill in any keys not present in pairs.
 */
export function parseOptions(
  pairs: readonly string[],
  specs: readonly OptionSpec[],
): Options {
  const specByName = new Map<string, OptionSpec>();
  for (const s of specs) {
    specByName.set(s.name, s);
  }

  const out: Record<string, string | number | boolean> = {};
  for (const raw of pairs) {
    const eq = raw.indexOf('=');
    const hasEq = eq !== -1;
    const key = (hasEq ? raw.slice(0, eq) : raw).trim();
    const val = hasEq ? raw.slice(eq + 1) : '';
    if (key === '') {
      throw new Error(`empty option key in "${raw}"`);
    }
    const spec = specByName.get(key);
    if (!spec) {
      const valid = specs.map(s => s.name).join(', ');
      throw new Error(`unknown option "${key}" (valid: ${valid})`);
    }
    if (!hasEq) {
      if (spec.type !== 'bool') {
        throw new Error(`option "${key}" requires a value (e.g. ${key}=...)`);
      }
      out[key] = true;
      continue;
    }
    out[key] = coerce(spec, val);
  }

  for (const s of specs) {
    if (Object.prototype.hasOwnProperty.call(out, s.name)) continue;
    if (s.default !== undefined) {
      out[s.name] = s.default;
    }
  }
  return out;
}

function coerce(
  spec: OptionSpec,
  val: string,
): string | number | boolean {
  switch (spec.type) {
    case 'string':
      return val;
    case 'int': {
      // Go's strconv.Atoi accepts only base-10, no leading/trailing space.
      if (!/^-?\d+$/.test(val)) {
        throw new Error(`option "${spec.name}": "${val}" is not an int`);
      }
      const n = Number(val);
      if (!Number.isSafeInteger(n)) {
        throw new Error(`option "${spec.name}": "${val}" is not an int`);
      }
      return n;
    }
    case 'bool': {
      // Mirrors Go's strconv.ParseBool accepted set.
      const truthy = new Set(['1', 't', 'T', 'TRUE', 'true', 'True']);
      const falsy = new Set(['0', 'f', 'F', 'FALSE', 'false', 'False']);
      if (truthy.has(val)) return true;
      if (falsy.has(val)) return false;
      throw new Error(`option "${spec.name}": "${val}" is not a bool`);
    }
    case 'enum': {
      const allowed = spec.enum ?? [];
      if (allowed.includes(val)) return val;
      throw new Error(
        `option "${spec.name}": "${val}" not in {${allowed.join(', ')}}`,
      );
    }
  }
}

/** Pretty-print an OptionType. */
export function optionTypeName(t: OptionType): string {
  return t;
}

File: go/console/output/formatter.go
package output

import (
	"fmt"
	"io"
	"strconv"
	"strings"
)

// Formatter encodes a value to an io.Writer in a specific format.
//
// Implementations declare their key, file extensions, and the option keys
// they accept. The registry validates --format-opt input against Options()
// before invoking Render, so Render can trust opts to contain only declared
// keys with values coerced to declared types.
type Formatter interface {
	// Key returns the unique format identifier exposed via --format <key>.
	Key() string

	// Extensions returns file extensions (with leading dot, e.g. ".csv")
	// that map to this formatter for --output extension inference.
	// May return an empty slice when no extension applies.
	Extensions() []string

	// Options returns the option specs accepted by this formatter via
	// --format-opt key=value. Returning nil means the formatter has no
	// options.
	Options() []OptionSpec

	// Render writes data to w. opts contains only validated option values
	// keyed by OptionSpec.Name; missing keys mean the caller did not set
	// them (Render should fall back to spec defaults via opts.GetOr).
	// cols is the user-requested column projection; an empty slice means
	// "all default columns".
	Render(w io.Writer, data any, opts Options, cols []string) error
}

// OptionType identifies the kind of value an OptionSpec accepts.
type OptionType int

const (
	OptString OptionType = iota
	OptInt
	OptBool
	OptEnum
)

// OptionSpec describes one option accepted by a Formatter.
type OptionSpec struct {
	Name    string     // option key, e.g. "delimiter"
	Type    OptionType // value kind
	Default any        // default value (must match Type); zero value if nil
	Usage   string     // short help text for --format-help
	Enum    []string   // allowed values when Type == OptEnum
}

// Options is a validated map of option values produced by parsing
// --format-opt key=value pairs against a Formatter's Options() specs.
type Options map[string]any

// GetString returns the string value for key or "" if absent.
func (o Options) GetString(key string) string {
	v, ok := o[key]
	if !ok {
		return ""
	}
	s, _ := v.(string)
	return s
}

// GetInt returns the int value for key or 0 if absent.
func (o Options) GetInt(key string) int {
	v, ok := o[key]
	if !ok {
		return 0
	}
	n, _ := v.(int)
	return n
}

// GetBool returns the bool value for key or false if absent.
func (o Options) GetBool(key string) bool {
	v, ok := o[key]
	if !ok {
		return false
	}
	b, _ := v.(bool)
	return b
}

// GetOr returns the value for key or fallback if absent.
func (o Options) GetOr(key string, fallback any) any {
	if v, ok := o[key]; ok {
		return v
	}
	return fallback
}

// ParseOptions validates raw key=value pairs against specs and returns the
// coerced Options map. Unknown keys, type errors, and out-of-enum values
// produce an error listing the offending key and the valid set.
//
// A pair with no '=' (e.g. "no-header") is treated as bool true; this is
// only valid when the matching spec has Type == OptBool.
//
// Defaults from specs are filled in for any keys not present in pairs.
func ParseOptions(pairs []string, specs []OptionSpec) (Options, error) {
	specByName := make(map[string]OptionSpec, len(specs))
	for _, s := range specs {
		specByName[s.Name] = s
	}

	out := Options{}
	for _, raw := range pairs {
		key, val, hasEq := strings.Cut(raw, "=")
		key = strings.TrimSpace(key)
		if key == "" {
			return nil, fmt.Errorf("empty option key in %q", raw)
		}
		spec, ok := specByName[key]
		if !ok {
			return nil, fmt.Errorf("unknown option %q (valid: %s)",
				key, strings.Join(specNames(specs), ", "))
		}
		if !hasEq {
			if spec.Type != OptBool {
				return nil, fmt.Errorf("option %q requires a value (e.g. %s=...)", key, key)
			}
			out[key] = true
			continue
		}
		coerced, err := coerce(spec, val)
		if err != nil {
			return nil, err
		}
		out[key] = coerced
	}

	for _, s := range specs {
		if _, ok := out[s.Name]; ok {
			continue
		}
		if s.Default != nil {
			out[s.Name] = s.Default
		}
	}
	return out, nil
}

func coerce(spec OptionSpec, val string) (any, error) {
	switch spec.Type {
	case OptString:
		return val, nil
	case OptInt:
		n, err := strconv.Atoi(val)
		if err != nil {
			return nil, fmt.Errorf("option %q: %q is not an int", spec.Name, val)
		}
		return n, nil
	case OptBool:
		b, err := strconv.ParseBool(val)
		if err != nil {
			return nil, fmt.Errorf("option %q: %q is not a bool", spec.Name, val)
		}
		return b, nil
	case OptEnum:
		for _, allowed := range spec.Enum {
			if val == allowed {
				return val, nil
			}
		}
		return nil, fmt.Errorf("option %q: %q not in {%s}",
			spec.Name, val, strings.Join(spec.Enum, ", "))
	default:
		return nil, fmt.Errorf("option %q: unknown type", spec.Name)
	}
}

func specNames(specs []OptionSpec) []string {
	names := make([]string, len(specs))
	for i, s := range specs {
		names[i] = s.Name
	}
	return names
}
